File: test/language/explicit_parameters/java/explicit_parameters/explicit_enum_param/ExplicitEnumParamTest.java
package explicit_parameters.explicit_enum_param;

import static org.junit.Assert.assertEquals;

import java.io.File;
import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import explicit_parameters.ExplicitParametersDb;

import zserio.runtime.ZserioError;

public class ExplicitEnumParamTest
{
    @BeforeClass
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Before
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new ExplicitParametersDb(file.toString());
        database.createSchema();
    }

    @After
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void readWithoutCondition() throws SQLException, IOException, ZserioError
    {
        final EnumParamTable enumParamTable = database.getEnumParamTable();

        final List<EnumParamTableRow> writtenRows = new ArrayList<EnumParamTableRow>();
        fillEnumParamTableRows(writtenRows);
        enumParamTable.write(writtenRows);

        final EnumParamTableParameterProvider parameterProvider = new EnumParamTableParameterProvider();
        final List<EnumParamTableRow> readRows = enumParamTable.read(parameterProvider);
        checkEnumParamTableRows(writtenRows, readRows);
    }

    @Test
    public void readWithCondition() throws SQLException, IOException, ZserioError
    {
        final EnumParamTable enumParamTable = database.getEnumParamTable();

        final List<EnumParamTableRow> writtenRows = new ArrayList<EnumParamTableRow>();
        fillEnumParamTableRows(writtenRows);
        enumParamTable.write(writtenRows);

        final EnumParamTableParameterProvider parameterProvider = new EnumParamTableParameterProvider();
        final String condition = "name='Name1'";
        final List<EnumParamTableRow> readRows = enumParamTable.read(parameterProvider, condition);
        assertEquals(1, readRows.size());

        final int expectedRowNum = 1;
        final EnumParamTableRow readRow = readRows.get(0);
        checkEnumParamTableRow(writtenRows.get(expectedRowNum), readRow);
    }

    @Test
    public void update() throws SQLException, IOException, ZserioError
    {
        final EnumParamTable enumParamTable = database.getEnumParamTable();

        final List<EnumParamTableRow> writtenRows = new ArrayList<EnumParamTableRow>();
        fillEnumParamTableRows(writtenRows);
        enumParamTable.write(writtenRows);

        final int updateRowId = 3;
        final EnumParamTableRow updateRow = createEnumParamTableRow(updateRowId, "UpdatedName");
        final String updateCondition = "id=" + updateRowId;
        enumParamTable.update(updateRow, updateCondition);

        final EnumParamTableParameterProvider parameterProvider = new EnumParamTableParameterProvider();
        final List<EnumParamTableRow> readRows = enumParamTable.read(parameterProvider, updateCondition);
        assertEquals(1, readRows.size());

        final EnumParamTableRow readRow = readRows.get(0);
        checkEnumParamTableRow(updateRow, readRow);
    }

    private static class EnumParamTableParameterProvider implements EnumParamTable.ParameterProvider
    {
        @Override
        public TestEnum getCount1(ResultSet resultSet)
        {
            return ENUM_PARAM_TABLE_COUNT1;
        }

        @Override
        public TestEnum getCount2(ResultSet resultSet)
        {
            return ENUM_PARAM_TABLE_COUNT2;
        }
    }

    private static void fillEnumParamTableRows(List<EnumParamTableRow> rows)
    {
        for (int id = 0; id < NUM_ENUM_PARAM_TABLE_ROWS; ++id)
        {
            rows.add(createEnumParamTableRow(id, "Name" + id));
        }
    }

    private static EnumParamTableRow createEnumParamTableRow(long id, String name)
    {
        final EnumParamTableRow row = new EnumParamTableRow();

        row.setId(id);
        row.setName(name);

        final byte[] values1 = new byte[ENUM_PARAM_TABLE_COUNT1.getValue()];
        for (int i = 0; i < values1.length; ++i)
            values1[i] = (byte)id;
        final TestBlob testBlob1 = new TestBlob(ENUM_PARAM_TABLE_COUNT1, values1);
        row.setBlob1(testBlob1);

        final byte[] values2 = new byte[ENUM_PARAM_TABLE_COUNT2.getValue()];
        for (int i = 0; i < values2.length; ++i)
            values2[i] = (byte)(id + 1);
        final TestBlob testBlob2 = new TestBlob(ENUM_PARAM_TABLE_COUNT2, values2);
        row.setBlob2(testBlob2);

        final byte[] values3 = new byte[ENUM_PARAM_TABLE_COUNT1.getValue()];
        for (int i = 0; i < values3.length; ++i)
            values3[i] = (byte)(id + 2);
        final TestBlob testBlob3 = new TestBlob(ENUM_PARAM_TABLE_COUNT1, values3);
        row.setBlob3(testBlob3);

        return row;
    }

    private static void checkEnumParamTableRows(List<EnumParamTableRow> rows1, List<EnumParamTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkEnumParamTableRow(rows1.get(i), rows2.get(i));
    }

    private static void checkEnumParamTableRow(EnumParamTableRow row1, EnumParamTableRow row2)
    {
        assertEquals(row1.getId(), row2.getId());
        assertEquals(row1.getName(), row2.getName());
        assertEquals(row1.getBlob1(), row2.getBlob1());
        assertEquals(row1.getBlob2(), row2.getBlob2());
        assertEquals(row1.getBlob3(), row2.getBlob3());

        // check reused explicit count1 parameter
        assertEquals(row2.getBlob1().getCount(), row2.getBlob3().getCount());
    }

    private static final int NUM_ENUM_PARAM_TABLE_ROWS = 5;
    private static final TestEnum ENUM_PARAM_TABLE_COUNT1 = TestEnum.TEN;
    private static final TestEnum ENUM_PARAM_TABLE_COUNT2 = TestEnum.ELEVEN;
    private static final String FILE_NAME = "explicit_enum_param_test.sqlite";

    private final File file = new File(FILE_NAME);
    private ExplicitParametersDb database = null;
}

File: test/extensions/language/explicit_parameters/python/ExplicitEnumParamTest.py
import os

import ExplicitParameters


class ExplicitEnumParamTest(ExplicitParameters.TestCaseWithDb):
    @classmethod
    def setUpClass(cls):
        super(ExplicitEnumParamTest, cls).setUpClass()

        test_api = cls.api.explicit_enum_param

        class EnumParamTableParameterProvider(test_api.EnumParamTable.IParameterProvider):
            @staticmethod
            def count1(_row):
                return test_api.TestEnum.TEN

            @staticmethod
            def count2(_row):
                return test_api.TestEnum.ELEVEN

        cls.EnumParamTableParameterProvider = EnumParamTableParameterProvider

    def setUp(self):
        if os.path.exists(self.dbFileName):
            os.remove(self.dbFileName)
        self._database = self.api.ExplicitParametersDb.from_file(self.dbFileName)
        self._database.create_schema()

        self._enumParamTableCount1 = self.api.explicit_enum_param.TestEnum.TEN
        self._enumParamTableCount2 = self.api.explicit_enum_param.TestEnum.ELEVEN

    def tearDown(self):
        self._database.close()

    def testDeleteTable(self):
        self.assertTrue(self._isTableInDb())

        testTable = self._database.enum_param_table
        testTable.delete_table()
        self.assertFalse(self._isTableInDb())

        testTable.create_table()
        self.assertTrue(self._isTableInDb())

    def testReadWithoutCondition(self):
        testTable = self._database.enum_param_table

        writtenRows = self._createEnumParamTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.EnumParamTableParameterProvider()
        readRows = testTable.read(parameterProvider)
        numReadRows = 0
        for readRow in readRows:
            self._checkEnumParamTableRow(writtenRows[numReadRows], readRow)
            numReadRows += 1
        self.assertTrue(len(writtenRows), numReadRows)

    def testReadWithCondition(self):
        testTable = self._database.enum_param_table

        writtenRows = self._createEnumParamTableRows()
        testTable.write(writtenRows)

        parameterProvider = self.EnumParamTableParameterProvider()
        condition = "name='Name1'"
        readRows = testTable.read(parameterProvider, condition)
        expectedRowNum = 1
        for readRow in readRows:
            self._checkEnumParamTableRow(writtenRows[expectedRowNum], readRow)

    def testUpdate(self):
        testTable = self._database.enum_param_table

        writtenRows = self._createEnumParamTableRows()
        testTable.write(writtenRows)

        updateRowId = 3
        updateRow = self._createEnumParamTableRow(updateRowId, "UpdatedName")
        updateCondition = "id=" + str(updateRowId)
        testTable.update(updateRow, updateCondition)

        parameterProvider = self.EnumParamTableParameterProvider()
        readRows = testTable.read(parameterProvider, updateCondition)
        for readRow in readRows:
            self._checkEnumParamTableRow(updateRow, readRow)

    def _createEnumParamTableRows(self):
        rows = []
        for rowId in range(NUM_ENUM_PARAM_TABLE_ROWS):
            rows.append(self._createEnumParamTableRow(rowId, "Name" + str(rowId)))

        return rows

    def _createEnumParamTableRow(self, rowId, name):
        values1 = [rowId for i in range(self._enumParamTableCount1.value)]
        testBlob1 = self.api.explicit_enum_param.TestBlob(self._enumParamTableCount1, values1)

        values2 = [rowId + 1 for i in range(self._enumParamTableCount2.value)]
        testBlob2 = self.api.explicit_enum_param.TestBlob(self._enumParamTableCount2, values2)

        values3 = [rowId + 2 for i in range(self._enumParamTableCount1.value)]
        testBlob3 = self.api.explicit_enum_param.TestBlob(self._enumParamTableCount1, values3)

        return (rowId, name, testBlob1, testBlob2, testBlob3)

    def _checkEnumParamTableRow(self, row1, row2):
        self.assertEqual(row1, row2)

        # check reused explicit count1 parameter
        blob1 = row2[2]
        blob3 = row2[4]
        self.assertEqual(blob1.count, blob3.count)

    def _isTableInDb(self):
        # check if database does contain table
        sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + TABLE_NAME + "'"
        for row in self._database.connection.cursor().execute(sqlQuery):
            if len(row) == 1 and row[0] == TABLE_NAME:
                return True

        return False


TABLE_NAME = "enumParamTable"
NUM_ENUM_PARAM_TABLE_ROWS = 5
